velox/functions/sparksql/Hash.cpp

```diff
@@ -26,16 +26,280 @@ namespace {
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+ReturnType hashOne(int64_t input, SeedType seed) {
+  return HashClass::hashInt64(input, seed);
+}
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+ReturnType hashOne(float input, SeedType seed) {
+  return HashClass::hashFloat(input, seed);
+}
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+ReturnType hashOne(double input, SeedType seed) {
+  return HashClass::hashDouble(input, seed);
+}
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+ReturnType hashOne(int128_t input, SeedType seed) {
+  return HashClass::hashLongDecimal(input, seed);
+}
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+ReturnType hashOne(Timestamp input, SeedType seed) {
+  return HashClass::hashTimestamp(input, seed);
+}
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+ReturnType hashOne(StringView input, SeedType seed) {
+  return HashClass::hashBytes(input, seed);
+}
+
+template <
+    typename HashClass,
+    TypeKind kind,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+class PrimitiveVectorHasher;
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+class ArrayVectorHasher;
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+class MapVectorHasher;
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+class RowVectorHasher;
+
+template <
+    typename HashClass,
+    typename SeedType = typename HashTraits<HashClass>::SeedType,
+    typename ReturnType = typename HashTraits<HashClass>::ReturnType>
+class VectorHasher {
+ public:
+  // Compute the hash value of input vector at index.
+  ReturnType hashAt(vector_size_t index, SeedType seed) {
+    if (decoded_.isNullAt(index)) {
+      return seed;
+    }
+    return hashNotNull(index, seed);
+  }
+
+  virtual ReturnType hashNotNull(vector_size_t index, SeedType seed) = 0;
+
+  VectorHasher(DecodedVector& decoded) : decoded_(decoded) {}
```

Comment:
Maybe better to move the constructor and deconstructor near the class name for readability.